ignite/metrics/psnr.py

```diff
@@ -0,0 +1,114 @@
+from typing import Callable, Optional, Sequence, Union
+
+import torch
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+
+__all__ = ["PSNR"]
+
+
+class PSNR(Metric):
+    r"""Computes average `Peak signal-to-noise ratio (PSNR) <https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio>`_.
+
+    .. math::
+        \text{PSNR}(i, j) = 10 * \log_{10}\left(\frac{ MAX_{i}^2 }{ \text{ MSE } }\right)
+
+    - `y_pred` and `y` must be in the following shape (batch_size, ...).
+    - `y_pred` and `y` must have same dtype and same shape.
+
+    Args:
+        data_range (int or float): Range of the image. If not provided, it will be determined
+            from the image type. Typically, ``1.0`` or ``255``.
+        output_transform (callable, optional): A callable that is used to transform the Engine’s
+            process_function’s output into the form expected by the metric.
+        device (str or torch.device): specifies which device updates are accumulated on.
+            Setting the metric’s device to be the same as your update arguments ensures
+            the update method is non-blocking. By default, CPU.
+
+    Example:
+
+    To use with ``Engine`` and ``process_function``, simply attach the metric instance to the engine.
+    The output of the engine's ``process_function`` needs to be in format of
+    ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y, ...}``.
+
+    .. code-block:: python
+
+        def process_function(engine, batch):
+            # ...
+            return y_pred, y
+        engine = Engine(process_function)
+        psnr = PSNR()
+        psnr.attach(engine, "psnr")
+
+    .. versionadded:: 0.5.0
+    """
+
+    def __init__(
+        self,
+        data_range: Optional[Union[int, float]] = None,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+    ):
+        super().__init__(output_transform=output_transform, device=device)
+        self.data_range = data_range
+        self.__data_range = data_range
```

Comment:
@vfdev-5 @sdesrozis Is there any other way to avoid this? What we need to check is
- if `data_range` is given, then we use given `data_range`
- if not, we need to determine `data_range` every batch

Reply:
you can use `self.data_range` as None if not defined by the user and in the update you can use a local var instead of `self.data_range`:
```python
data_range = self.data_range
if data_range is None:
     # compute it for the batch

data_range_sqrt = data_range ** 2 
```